README.pod
lib/Plack/App/WebSocket.pm

Changes
Revision history for Plack-App-WebSocket

0.04    2014-08-17
        [PACKAGING]
        - Migrate to Module::Build::Prereqs::FromCPANfile.

0.03    2014-02-05
        [ENHANCEMENT]
        - (gh#1) Now on_established callback gets $psgi_env as the second
          argument. That way, you can change behavior of the connection
          according to query params etc. Thanks leedo.

0.02    2013-11-21
        [BUGFIX]
        - Sometimes installation failed because $VERSION was inspected
          before prerequisites were installed. This bug is fixed now
          that $VERSIONs are set by literals.

0.01    2013-11-10
        First version, released on an unsuspecting world.



t/testlib/Echo.pm
package testlib::Echo;
use strict;
use warnings;
use Test::More;
use testlib::Util qw(set_timeout run_server);
use utf8;
use Encode;

use AnyEvent::WebSocket::Client;
use AnyEvent;

use Plack::App::WebSocket;

sub run_tests {
    my ($server_runner) = @_;
    set_timeout;
    my $app = Plack::App::WebSocket->new(on_establish => sub {
        my $conn = shift;
        note("server established.");
        isa_ok($conn, "Plack::App::WebSocket::Connection");
        $conn->on(message => sub {
            my ($conn, $msg) = @_;
            note("server received message.");
            $conn->send($msg);
        });
        $conn->on(finish => sub {
            note("server finished.");
            undef $conn;
        });
        $conn->send("echo started");
    });
    my ($port, $server_guard) = run_server($server_runner, $app->to_app);

    my @test_data = (
        {label => "8 bytes", data => "AAAABBBB"},
        {label => "0 bytes", data => ""},
        {label => "zero", data => 0},
        {label => "256 bytes", data => "A" x 256},
        {label => "64 ki bytes", data => "A" x (64 * 1024)},
        {label => "UTF-8 encoded text", data => Encode::encode("utf8", 'ＵＴＦー８ ＴＥＸＴ')},
    );

    my $client = AnyEvent::WebSocket::Client->new;
    note("--- create new connection for each test");
    foreach my $test (@test_data) {
        my $conn = $client->connect("ws://127.0.0.1:$port/")->recv;
        note("client established.");
        my $cv_fin = AnyEvent->condvar;
        my @messages = ();
        $conn->on(each_message => sub {
            note("client received message.");
            push(@messages, $_[1]->body);
        });
        $conn->on(finish => sub {
            note("client finished.");
            $cv_fin->send;
        });
        $conn->send($test->{data});
        $conn->close;
        $cv_fin->recv;
        is_deeply(\@messages, ["echo started", $test->{data}], "new conn: $test->{label} OK");
    }

    {
        note("--- all tests by single connection");
        my $conn = $client->connect("ws://127.0.0.1:$port/")->recv;
        my @exp_message = ({label => "sent by sever", exp => "echo started"});
        my $cv_fin = AnyEvent->condvar;
        $conn->on(each_message => sub {
            my ($conn, $msg) = @_;
            my $exp = shift @exp_message;
            is($msg->body, $exp->{exp}, "$exp->{label}: server message OK");
        });
        $conn->on(finish => sub { $cv_fin->send });
        foreach my $test (@test_data) {
            push(@exp_message, {label => $test->{label}, exp => $test->{data}});
            $conn->send($test->{data});
        }
        $conn->close;
        $cv_fin->recv;
        is(scalar(@exp_message), 0, "expected messages are all received.");
    }
}

1;


t/testlib/Util.pm
package testlib::Util;
use strict;
use warnings;
use AnyEvent;
use Exporter qw(import);
use Test::More;
use Net::EmptyPort qw(empty_port);  ## should it be in Test::Requires?

our @EXPORT_OK = qw(set_timeout run_server);

my $TIMEOUT_DEFAULT_SEC = 10;

sub set_timeout {
    my $w; $w = AnyEvent->timer(after => $TIMEOUT_DEFAULT_SEC, cb => sub {
        undef $w;
        fail("Timeout");
        exit 2;
    });
}

sub run_server {
    my ($server_runner, $app) = @_;
    my $port = empty_port();
    return ($port, $server_runner->($port, $app));
}

1;


t/testlib/Cycle.pm
package testlib::Cycle;
use strict;
use warnings;
use Test::More;
use Test::Builder;
use Plack::App::WebSocket;
use AnyEvent::WebSocket::Client;
use testlib::Util qw(set_timeout run_server);
use AnyEvent;
use Scalar::Util qw(weaken);

sub run_tests {
    my ($server_runner) = @_;
    set_timeout;
    my $server_conn;
    my @messages = ();
    my $cv_finish = AnyEvent->condvar;
    my $cv_establish = AnyEvent->condvar;
    my $app = Plack::App::WebSocket->new(on_establish => sub {
        my $conn = shift;
        if(defined($server_conn)) {
            fail("More than one connection. something is wrong.");
            return;
        }
        $cv_finish->begin;
        $server_conn = $conn;
        weaken $server_conn;
        $cv_establish->send;
        $conn->on(message => sub {
            my ($conn, $message) = @_;
            push(@messages, $message);
            if($message eq "quit") {
                $conn->close();
            }
        });
        $conn->on(finish => sub {
            undef $conn;
            $cv_finish->end;
        });
    });
    
    my ($port, $guard) = run_server($server_runner, $app->to_app);
    my $client = AnyEvent::WebSocket::Client->new;
    my $client_conn = $client->connect("ws://127.0.0.1:$port/")->recv;
    $cv_establish->recv;
    ok defined($server_conn), "server connection is alive because of the looping closure at finish callback";
    $cv_finish->begin;
    $client_conn->on(finish => sub { undef $client_conn; $cv_finish->end });
    $client_conn->send("hoge");
    $client_conn->send("quit");
    $cv_finish->recv;
    is_deeply \@messages, ["hoge", "quit"], "messages received by the server OK";
    ok !defined($server_conn), "server connection is destroyed because of the self-destruction in finish callback";
}

1;


lib/Plack/App/WebSocket.pm
package Plack::App::WebSocket;
use strict;
use warnings;
use parent qw(Plack::Component);
use Carp;
use Plack::Response;
use AnyEvent::WebSocket::Server;
use Try::Tiny;
use Plack::App::WebSocket::Connection;

our $VERSION = "0.04";

my $ERROR_ENV = "plack.app.websocket.error";

sub new {
    my ($class, @args) = @_;
    my $self = $class->SUPER::new(@args);
    croak "on_establish param is mandatory" if not defined $self->{on_establish};
    croak "on_establish param must be a code-ref" if ref($self->{on_establish}) ne "CODE";
    $self->{on_error} ||= \&_default_on_error;
    croak "on_error param must be a code-ref" if ref($self->{on_error}) ne "CODE";
    
    $self->{websocket_server} = AnyEvent::WebSocket::Server->new();
    return $self;
}

sub _default_on_error {
    my ($env) = @_;
    my $res = Plack::Response->new;
    $res->content_type("text/plain");
    if(!defined($env->{$ERROR_ENV})) {
        $res->status(500);
        $res->body("Unknown error");
    }elsif($env->{$ERROR_ENV} eq "not supported by the PSGI server") {
        $res->status(500);
        $res->body("The server does not support WebSocket.");
    }elsif($env->{$ERROR_ENV} eq "invalid request") {
        $res->status(400);
        $res->body("The request is invalid for a WebSocket request.");
    }else {
        $res->status(500);
        $res->body("Unknown error: $env->{$ERROR_ENV}");
    }
    $res->content_length(length($res->body));
    return $res->finalize;
}

sub _respond_via {
    my ($responder, $psgi_res) = @_;
    if(ref($psgi_res) eq "CODE") {
        $psgi_res->($responder);
    }else {
        $responder->($psgi_res);
    }
}

sub call {
    my ($self, $env) = @_;
    if(!$env->{"psgi.streaming"} || !$env->{"psgi.nonblocking"} || !$env->{"psgix.io"}) {
        $env->{$ERROR_ENV} = "not supported by the PSGI server";
        return $self->{on_error}->($env);
    }
    my $cv_conn = $self->{websocket_server}->establish_psgi($env, $env->{"psgix.io"});
    return sub {
        my $responder = shift;
        $cv_conn->cb(sub {
            my ($cv_conn) = @_;
            my ($conn) = try { $cv_conn->recv };
            if(!$conn) {
                $env->{$ERROR_ENV} = "invalid request";
                _respond_via($responder, $self->{on_error}->($env));
                return;
            }
            $self->{on_establish}->(Plack::App::WebSocket::Connection->new($conn, $responder), $env);
        });
    };
}

1;

__END__

=pod

=head1 NAME

Plack::App::WebSocket - WebSocket server as a PSGI application

=head1 SYNOPSIS

    use Plack::App::WebSocket;
    use Plack::Builder;
    
    builder {
        mount "/websocket" => Plack::App::WebSocket->new(
            on_error => sub {
                my $env = shift;
                return [500,
                        ["Content-Type" => "text/plain"],
                        ["Error: " . $env->{"plack.app.websocket.error"}]];
            },
            on_establish => sub {
                my $conn = shift; ## Plack::App::WebSocket::Connection object
                my $env = shift;  ## PSGI env
                $conn->on(
                    message => sub {
                        my ($conn, $msg) = @_;
                        $conn->send($msg);
                    },
                    finish => sub {
                        undef $conn;
                        warn "Bye!!\n";
                    },
                );
            }
        )->to_app;
        
        mount "/" => $your_app;
    };

=head1 DESCRIPTION

This module is a L<PSGI> application that creates an endpoint for WebSocket connections.

=head2 Prerequisites

To use L<Plack::App::WebSocket>, your L<PSGI> server must meet the following requirements.
(L<Twiggy> meets all of them, for example)

=over

=item *

C<psgi.streaming> environment is true.

=item *

C<psgi.nonblocking> environment is true, and the server supports L<AnyEvent>.

=item *

C<psgix.io> environment holds a valid raw IO socket object. See L<PSGI::Extensions>.

=back

=head1 CLASS METHODS

=head2 $app = Plack::App::WebSocket->new(%args)

The constructor.

Fields in C<%args> are:

=over

=item C<on_establish> => CODE (mandatory)

A subroutine reference that is called each time it establishes a new
WebSocket connection to a client.

The code is called like

    $code->($connection, $psgi_env)

where C<$connection> is a L<Plack::App::WebSocket::Connection> object
and C<$psgi_env> is the PSGI environment object for the connection request.
You can use the C<$connection> to communicate with the client.

Make sure you keep C<$connection> object as long as you need it.
If you lose reference to C<$connection> object and it's destroyed,
the WebSocket connection (and its underlying transport connection) is closed.

=item C<on_error> => PSGI_APP (optional)

A subroutine reference that is called when some error
happens while processing a request.

The code is a L<PSGI> app, so it's called like

    $psgi_response = $code->($psgi_env)

C<$psgi_response> is returned to the client instead of a valid
WebSocket handshake response.

When C<$code> is called, C<< $psgi_env->{"plack.app.websocket.error"} >> contains a string
that briefly describes the error (See below).

By default, it returns a simple non-200 HTTP response according to C<< $psgi_env->{"plack.app.websocket.error"} >>.
See below for detail.

=back

=head1 C<plack.app.websocket.error> ENVIRONMENT STRINGS

Below is the list of possible values of C<plack.app.websocket.error> L<PSGI> environment parameter.
It is set in the C<on_error> callback.

=over

=item C<"not supported by the PSGI server">

The L<PSGI> server does not support L<Plack::App::WebSocket>. See L</Prerequisites>.

By default, 500 "Internal Server Error" response is returned for this error.

=item C<"invalid request">

The client sent an invalid request.

By default, 400 "Bad Request" response is returned for this error.

=back

=head1 OBJECT METHODS

=head2 $psgi_response = $app->call($psgi_env)

Process the L<PSGI> environment (C<$psgi_env>) and returns a L<PSGI> response (C<$psgi_response>).

=head2 $app_code = $app->to_app

Return a L<PSGI> application subroutine reference.

=head1 SEE ALSO

=over

=item L<Amon2::Plugin::Web::WebSocket>

WebSocket implementation for L<Amon2> Web application framework.

=item L<Mojo::Transaction::WebSocket>

WebSocket implementation for L<Mojolicious> Web application framework.

=item L<PocketIO>

Socket.io implementation as a L<PSGI> application.

=item L<SockJS>

SockJS implementation as a L<PSGI> application.

=back

=head1 AUTHOR

Toshio Ito, C<< <toshioito at cpan.org> >>

=head1 CONTRIBUTORS

leedo

=head1 REPOSITORY

L<https://github.com/debug-ito/Plack-App-WebSocket>

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Toshio Ito.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


lib/Plack/App/WebSocket/Connection.pm
package Plack::App::WebSocket::Connection;
use strict;
use warnings;
use Carp;
use Scalar::Util qw(weaken);
use Devel::GlobalDestruction ();
use AnyEvent;

our $VERSION = "0.04";

sub new {
    my ($class, $conn, $responder) = @_;
    my $self = bless {
        connection => $conn,
        responder => $responder,
        handlers => {
            message => [],
            finish  => [],
        },
    }, $class;
    $self->_setup_internal_event_handlers();
    return $self;
}

sub _setup_internal_event_handlers {
    my ($self) = @_;
    weaken $self;
    $self->{connection}->on(each_message => sub {
        return if !defined($self);
        my $strong_self = $self; ## make sure $self is alive during callback execution
        $_->($self, $_[1]->body) foreach @{$self->{handlers}{message}};
    });
    $self->{connection}->on(finish => sub {
        return if !defined($self);
        my $strong_self = $self; ## make sure $self is alive during callback execution
        $_->($self) foreach @{$self->{handlers}{finish}};
    });
}

sub _clear_event_handlers {
    my ($self) = @_;
    foreach my $handler_list (values %{$self->{handlers}}) {
        @$handler_list = ();
    }
}

sub on {
    my ($self, %handlers) = @_;
    foreach my $event (keys %handlers) {
        my $handler = $handlers{$event};
        croak "handler for event $event must be a code-ref" if ref($handler) ne "CODE";
        $event = "finish" if $event eq "close";
        my $handler_list = $self->{handlers}{$event};
        croak "Unknown event: $event" if not defined $handler_list;
        push(@$handler_list, $handler);
    }
}

sub send {
    my ($self, $message) = @_;
    $self->{connection}->send($message);
}

sub close {
    my ($self) = @_;
    $self->{connection}->close;
}

our $WAIT_FOR_FLUSHING_SEC = 5;

sub DESTROY {
    my ($self) = @_;
    return if Devel::GlobalDestruction::in_global_destruction;
    $self->_clear_event_handlers();
    my $connection = $self->{connection};
    $connection->close();  ## explicit close because $responder may keep the socket.
    my $responder = $self->{responder};
    my $w; $w = AnyEvent->timer(after => $WAIT_FOR_FLUSHING_SEC, cb => sub {
        $responder->([200, ["Content-Type", "text/plain"], ["WebSocket finished"]]);
        undef $w;
        undef $responder;

        ## Prolong $connection's life as long as $responder. This is
        ## necessary to make sure $connection actively shuts down the
        ## socket. If $connection is destroyed immediately and the
        ## kernel's write buffer is full, $connection may fail to shut
        ## down the socket (because $connection delays the active
        ## shutdown after sending all the buffered data). If that
        ## happens, the socket stays open, which is bad.
        undef $connection;
    });
}

1;

__END__

=pod

=head1 NAME

Plack::App::WebSocket::Connection - WebSocket connection for Plack::App::WebSocket

=head1 SYNOPSIS

    my $app = Plack::App::WebSocket->new(on_establish => sub {
        my $connection = shift;
        $connection->on(message => sub {
            my ($connection, $message) = @_;
            warn "Received: $message\n";
            if($message eq "quit") {
                $connection->close();
            }
        });
        $connection->on(finish => sub {
            warn "Closed\n";
            undef $connection;
        });
        $connection->send("Message to the client");
    });

=head1 DESCRIPTION

L<Plack::App::WebSocket::Connection> is an object representing a
WebSocket connection to a client. It is created by
L<Plack::App::WebSocket> internally and given to you in
C<on_establish> callback function.

=head1 OBJECT METHODS

=head2 $connection->on($event => $handler)

Register a callback function to a particular event.
You can register multiple callbacks to the same event.

C<$event> is a string and C<$handler> is a subroutine reference.

Possible value for C<$event> is:

=over

=item C<"message">

    $handler->($connection, $message)

C<$handler> is called for each message received via the C<$connection>.
Argument C<$connection> is the L<Plack::App::WebSocket::Connection> object,
and C<$message> is a non-decoded byte string of the received message.

=item C<"finish"> (alias: C<"close">)

    $handler->($connection)

C<$handler> is called when the C<$connection> is closed.
Argument C<$connection> is the L<Plack::App::WebSocket::Connection> object.

=back

=head2 $connection->send($message)

Send a message via C<$connection>.

C<$message> should be a UTF-8 encoded string.

=head2 $connection->close()

Close the WebSocket C<$connection>.

=head1 AUTHOR

Toshio Ito, C<< <toshioito at cpan.org> >>


=cut
